x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java

```diff
@@ -43,55 +46,108 @@
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            (canPushToSource(exp, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats())) ? pushable : nonPushable).add(exp);
+        }
+        return rewrite(filterExec, queryExec, pushable, nonPushable);
+    }
+
+    private static PhysicalPlan planFilterExec(
+        FilterExec filterExec,
+        EvalExec evalExec,
+        EsQueryExec queryExec,
+        LocalPhysicalOptimizerContext ctx
+    ) {
+        LinkedHashMap<NameId, FieldAttribute> refs = new LinkedHashMap<>();
+        evalExec.fields().forEach(alias -> {
+            if (alias.child() instanceof FieldAttribute fieldAttribute) {
+                refs.put(alias.id(), fieldAttribute);
+            } else if (alias.child() instanceof ReferenceAttribute ref && refs.containsKey(ref.id())) {
+                FieldAttribute fieldAttribute = refs.get(ref.id());
+                refs.put(alias.id(), fieldAttribute);
             }
-            // Combine GT, GTE, LT and LTE in pushable to Range if possible
-            List<Expression> newPushable = combineEligiblePushableToRange(pushable);
-            if (newPushable.size() > 0) { // update the executable with pushable conditions
-                Query queryDSL = PlannerUtils.TRANSLATOR_HANDLER.asQuery(Predicates.combineAnd(newPushable));
-                QueryBuilder planQuery = queryDSL.asBuilder();
-                var query = Queries.combine(Queries.Clause.FILTER, asList(queryExec.query(), planQuery));
-                queryExec = new EsQueryExec(
-                    queryExec.source(),
-                    queryExec.index(),
-                    queryExec.indexMode(),
-                    queryExec.output(),
-                    query,
-                    queryExec.limit(),
-                    queryExec.sorts(),
-                    queryExec.estimatedRowSize()
-                );
-                if (nonPushable.size() > 0) { // update filter with remaining non-pushable conditions
-                    plan = new FilterExec(filterExec.source(), queryExec, Predicates.combineAnd(nonPushable));
-                } else { // prune Filter entirely
-                    plan = queryExec;
-                }
-            } // else: nothing changes
+        });
+        List<Expression> pushable = new ArrayList<>();
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            (canPushToSource(exp, refs, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats())) ? pushable : nonPushable).add(
+                exp
+            );
         }
+        // Replace field references with their actual field attributes
+        pushable.replaceAll(e -> e.transformDown(ReferenceAttribute.class, r -> fieldAttribute(r, refs)));
+        return rewrite(filterExec, queryExec, pushable, nonPushable);
```

Comment:
I don't think this can be right.

This disregards the EvalExec altogether. But the EvalExec might evaluate values that are used downstream, or it might evaluate non-pushable predicates for the FilterExec.

Reply:
I believe this is resolved in https://github.com/elastic/elasticsearch/pull/112938/commits/27a8b8cf62dad23fbd825a0bed81a7e88f1317c3. I maintain also the other aliases that need to be maintained in the EvalExec.